flutter_searchbox/lib/state_provider.dart

```diff
@@ -0,0 +1,255 @@
+  final dynamic error;
+  final dynamic value;
+  final List<Map<dynamic, dynamic>>? query;
+  final dynamic dataField;
+  final int? size;
+  final int? from;
+  final dynamic fuzziness;
+  final List<String>? includeFields;
+  final List<String>? excludeFields;
+  final SortType? sortBy;
+  final Map<String, dynamic>? react;
+  final Map Function(SearchController searchController)? defaultQuery;
+  final Map Function(SearchController searchController)? customQuery;
+
+  SearchControllerState({
+    this.aggregationData,
+    this.results,
+    this.requestStatus,
+    this.error,
+    this.value,
+    this.query,
+    this.dataField,
+    this.size,
+    this.from,
+    this.fuzziness,
+    this.includeFields,
+    this.excludeFields,
+    this.sortBy,
+    this.react,
+    this.defaultQuery,
+    this.customQuery,
+  });
+}
+
+/// It allows you to access the current state of your widgets along with the search results.
+/// For instance, you can use this component to create results/no results or query/no query pages.
+///
+/// Examples Use(s):
+///    - perform side-effects based on the results states of various widgets.
+///    - render custom UI based on the current state of app.
+class StateProvider extends StatefulWidget {
+  /// A map of widget ids and list of properties to subscribe to.
+  /// [Optional]
+  /// This property allows users to select the widgets for which
+  /// they want to listen to the state changes.
+  /// For example,
+  /// ```dart
+  /// StateProvider(
+  ///   ...
+  ///   subscribeTo: {
+  ///     'result-component': [KeysToSubscribe.Results, KeysToSubscribe.From]
+  ///   },
+  ///   ...
+  /// )
+  /// ```
+  final Map<String, List<KeysToSubscribe>>? subscribeTo;
+
+  /// Callback function,  is a callback function called when the search state changes
+  /// and accepts the previous and next search state as arguments.
+  /// For example,
+  /// ```dart
+  /// StateProvider(
+  ///   ...
+  ///   onChange: (prevState, nextState) {
+  ///         // do something here
+  ///     },
+  ///   ...
+  /// )
+  /// ```
+  final void Function(Map<String, SearchControllerState>,
+      Map<String, SearchControllerState>)? onChange;
+
+  /// It is used for rendering a custom UI based on updated state
+  /// For example,
+  /// ```dart
+  /// StateProvider(
+  ///   ...
+  ///   build: (controllerState) {
+  ///     return Text(
+  ///       'Total results on screen--- ${controllerState["result-widget"]?.results?.data?.length ?? ''} ',
+  ///       style: TextStyle(
+  ///         fontSize: 19.0,
+  ///         color: Colors.red,
+  ///       ),
+  ///     );
+  ///   },
+  ///   ...
+  /// )
+  /// ```
+  final Widget Function(Map<String, SearchControllerState>)? build;
+
+  StateProvider({
+    this.subscribeTo,
+    this.onChange,
+    this.build,
+    Key? key,
+  }) : super(key: key) {
+    if (build == null && onChange == null) {
+      throw ("Atleast one, build or onChange prop is required.");
```

Comment:
Use `assert`.
https://stackoverflow.com/questions/57833399/whats-the-recommended-way-in-dart-asserts-or-throw-errors

Reply:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/57627350/164389440-c5e0192d-d19a-4c6c-97f0-a22ad3bcf8bb.png">
is this good enough?